content/post/asdf.md

```diff
@@ -0,0 +1,250 @@
+3. The schema should be optional. Formats that require an extra schema file in
+a separate file make introspection harder while requiring ML researchers to
+maintain an extra piece of code. Pythonistas expect introspection, so for
+an idiomatic Python library this is essential.
+4. The typical data size may span over tens of gigabytes. Compared to gigabytes,
+the overhead of including a schema looks irrelevant indeed, hence the files
+become self-descriptive and the schema turns into a validation perk.
+This requirement rules out Protocol Buffers completely, because each PB file
+is always fully loaded into memory in all the existing implementations.
+The same requirement excludes popular binary serialization formats which target
+messaging and RPC and are not optimized for large messages with few items.
+1. Python should be a first-class citizen. numpy arrays should be serializable
+without any additional code.
+6. Yet the format should not require Python. The user-facing applications at source{d}
+are written in Go and otherwise it will be hard to integrate.
+7. On-the-fly compression. NLP models often contain strings which can require much
+space while being perfectly compressible. Integers can be compressed too since
+we sometimes don't know their range beforehand and use 32 bits while only 16 are
+really needed.
+
+[HDF5](https://support.hdfgroup.org/HDF5/) was the closest to those. It is binary,
+there is no schema, it supports big tensors, Python bindings are mature and
+well-integrated, there are bindings for other languages. HDF5 is used in e.g. [Keras](https://github.com/keras-team/keras).
+However, it is not ideal in terms of performance and lacks some modern features.
+Read ["Moving away from HDF5"](https://cyrille.rossant.net/moving-away-hdf5/).
+
+We previously had some positive experience with SQLite + SQLAlchemy on top,
+but of course that variant does not stand big data blobs (4).
+There is a common workaround for (4): it is possible to store huge tensors as external files.
+This implies concatenating all the blocks together before uploading and splitting
+them back after downloading, e.g. as a TAR or a ZIP without compression.
+Those meta-archives are typically used as a temporary transfer medium
+(e.g., [TensorFlow Hub](https://www.tensorflow.org/hub/hosting)).  
+In turn, this means the higher usage complexity, the doubled requirement for free disk size,
+the increased vulnerability to data corruptions and the explosion on the number
+of open file descriptors.
+
+No more intrigue: we discovered ASDF.
+
+## ASDF
+
+[Advanced Scientific Data Format](https://github.com/spacetelescope/asdf) (ASDF)
+is a next generation serialization format for scientific data. This means that
+it focuses on storing sparse and dense tensors in an efficient way.
+The ASDF project started by [Michael Droettboom](https://github.com/mdboom) (Matplotlib; astropy)
```

Comment:
@perrygreenfield deserves at least as much credit here as Mike: he has been the prime advocate for a new format and has been deeply involved in the design of the standard. I am still probably understating his involvement here, but I wasn't around at the time of ASDF's birth. It's probably be best to cite all of the [original paper](https://www.sciencedirect.com/science/article/pii/S2213133715000645?via%3Dihub) authors here.